docs/configuration.rst
Configuration
=============
The design of ``lorax`` is necessarily complex, with configuration occurring
in multiple layers at different times. Proceeding in order from highest to
lowest priority, these are:

======================= ==== ============================================================================
Layer                   Code            Description
======================= ==== ============================================================================
Program                 P    Some variables are output-only and defined (or overidden) at runtime.

Environmental Variable  E    Environmental variables beginning with ``LORAX_`` are parsed according to
                             the types of existing (lower-level) variables with matching name.  The
                             environmental variable names should be entirely upper-case.  If the variable
                             was not previously defined, it will be assumed to be of string type.

Config File             C    A config file for the instance (in the etc/ directory with prefix given by
                             the environmental variable LORAX_SETTINGS, default is ``lorax`` and extension
                             '.conf') is parsed before environmental variables.  The type is defined by
                             python parsing rules.  These values are changeable by the ``lorax config``
                             command.

Defaults                D    Default values are initialized depending on the value of the environmental
                             variable ``LORAX_CONFIGURATION``.  The following values are recognized:
                             * ``default``: a value suitable for testing and deployment
                             * ``development``: a debugging environment with much exposed.
                             * ``serverOnly``: same as default, but the queues are not started.
                             * ``treebuilder``: treebuilder queue is the only service started.
                             * ``aligner``: aligner queue is the only service started.

====================== ===== ============================================================================

Configuration Variables
-----------------------

======================= ===== ============================================================================
Configuration Variable  Code  Description
======================= ===== ============================================================================
ALERTMANAGER_DIR        P     Directory in which the alertmanager executable is found.

ALIGNERS                D     Dictionary of available aligners.

ALIGNMENT_QUEUE         D     Name of the alignment queue.

ALIGNMENT_QUEUE_TIMEOUT D     Maximum time-to-live for alignment jobs (default 86400)

CRASHMAIL_EMAIL         D     Email alert address.  Defaults to username.

CRASHMAIL_EVENTS        D     Defaults to "PROCESS_STATE_EXITED".

CURL_ARGS               P     Arguments for use with curl.

  DATA type(str) =  "/home/localhost/joelb/.lorax/0.94/var/data/"

  DATETIME type(str) =  "2017-09-19 13:19:07"

  DEBUG type(bool) =  False

  DIR_MODE type(str) =  "755"

  ENVIRONMENT_DUMP type(bool) =  False

  EXPLAIN_TEMPLATE_LOADING type(bool) =  False

  FASTTREE_EXE type(str) =  "FastTree"

  FILE_LOG_FORMAT type(str) =  "%(levelname)s: %(message)s"

  GUNICORN_LOG_LEVEL type(str) =  "debug"

  GUNICORN_UNIX_SOCKET type(bool) =  True

  GUNICORN_URL type(str) =  "unix://%(ENV_LORAX_VAR)s/run/gunicorn.sock"

  HMMALIGN_EXE type(str) =  "hmmalign"

  HOST type(str) =  "localhost"

  HOSTNAME type(str)      D   Hostname for nginx to use.  Defaults to

  JSONIFY_MIMETYPE type(str) =  "application/json"

  JSONIFY_PRETTYPRINT_REGULAR type(bool) =  True

  JSON_AS_ASCII type(bool) =  True

  JSON_SORT_KEYS type(bool) =  True
  LOG type(str) =  "/home/localhost/joelb/.lorax/0.94/var/log"
  LOGFILE type(bool) =  True
  LOGFILE_BACKUPCOUNT type(int) =  1
  LOGFILE_MAXBYTES type(int) =  10000000
  LOGFILE_NAME type(NoneType) =  None
  LOGGER_HANDLER_POLICY type(str) =  "always"
  LOGGER_NAME type(str) =  "lorax"
  MAX_CONTENT_LENGTH type(NoneType) =  None
  NAME type(str) =  "lorax"
  NGINX_EVENTS type(str) =  "use epoll;"
  NGINX_LISTEN_ARGS type(str) =  "deferred"
  NGINX_SERVER_NAME type(str) =  "localhost"
  NGINX_UNIX_SOCKET type(bool) =  False
  NGINX_URL type(str) =  "localhost:58927"
  NODE_EXPORTER_DIR type(str) =  "/home/localhost/joelb/.lorax/0.94/node_exporter-0.14.0.linux-amd64"
  PERMANENT_SESSION_LIFETIME type(timedelta) =  31 days, 0:00:00
  PLATFORM type(str) =  "Linux"
  PREFERRED_URL_SCHEME type(str) =  "http"
  PRESERVE_CONTEXT_ON_EXCEPTION type(NoneType) =  None
  PROCESS_UMASK type(str) =  "0002"
  PROJECT_HOME type(str) =  "https://github.com/LegumeFederation/lorax"
  PROMETHEUS_DIR type(str) =  "/home/localhost/joelb/.lorax/0.94/prometheus-2.0.0-beta.2.linux-amd64"
  PROPAGATE_EXCEPTIONS type(NoneType) =  None
  PUSH_GATEWAY_DIR type(NoneType) =  None
  QUIET type(bool) =  False
  RAXML_EXE type(str) =  "raxmlHPC"
  RC_GROUP type(str) =  ""
  RC_USER type(str) =  "joelb"
  RC_VERBOSE type(bool) =  False
  REDIS_DB type(int) =  0
  REDIS_HOST type(str) =  "localhost"
  REDIS_PASSWORD type(NoneType) =  None
  REDIS_PORT type(int) =  0
  REDIS_UNIX_SOCKET type(bool) =  True
  REDIS_URL type(str) =  "unix://@'/home/localhost/joelb/.lorax/0.94/var/run/redis.sock?db=0"
  RQ_ASYNC type(bool) =  True
  RQ_JOB_CLASS type(str) =  "rq.job.Job"
  RQ_POLL_INTERVAL type(int) =  2500
  RQ_QUEUE_CLASS type(str) =  "rq.queue.Queue"
  RQ_REDIS_HOST type(str) =  "localhost"
  RQ_REDIS_PORT type(int) =  0
  RQ_REDIS_URL type(str) =  "unix://@'/home/localhost/joelb/.lorax/0.94/var/run/redis.sock?db=0"
  RQ_SCHEDULER_INTERVAL type(int) =  60
  RQ_SCHEDULER_QUEUE type(str) =  "alignment"
  RQ_UNIXSOCKET type(str) =  "unixsocket /home/localhost/joelb/.lorax/0.94/var/run/redis.sock"
  RQ_WORKER_CLASS type(str) =  "rq.worker.Worker"
  SEND_FILE_MAX_AGE_DEFAULT type(timedelta) =  12:00:00
  SERVER_NAME type(NoneType) =  None
  SESSION_COOKIE_DOMAIN type(NoneType) =  None
  SESSION_COOKIE_HTTPONLY type(bool) =  True
  SESSION_COOKIE_NAME type(str) =  "session"
  SESSION_COOKIE_PATH type(NoneType) =  None
  SESSION_COOKIE_SECURE type(bool) =  False
  SESSION_REFRESH_EACH_REQUEST type(bool) =  True
  SETTINGS type(str) =  "lorax.conf"
  STDERR_LOG_FORMAT type(str) =  "%(levelname)s: %(message)s"
  TEMPLATES_AUTO_RELOAD type(NoneType) =  None
  TESTING type(bool) =  False
  THREADS type(int) =  0
  TMP type(str) =  "/home/localhost/joelb/.lorax/0.94/var/tmp"
  TRAP_BAD_REQUEST_ERRORS type(bool) =  False
  TRAP_HTTP_EXCEPTIONS type(bool) =  False
  TREEBUILDERS type(dict) =  {'FastTree': {'peptide': ['-nopr', '-log', 'peptide.log'], 'DNA': ['-nt', '-gtr', '-log', 'nucleotide.log', '-nopr']}, 'RAxML': {'peptide': ['-b', '12345', '-p', '12345', '-N', '10', '-m', 'PROTGAMMABLOSUM62'], 'DNA': ['-d']}}
  TREE_QUEUE type(str) =  "treebuilding"
  TREE_QUEUE_TIMEOUT type(int) =  2592000
  URL type(str) =  "http://localhost:58927"
  USER_CONFIG_PATH type(str) =  "~/.lorax"
  USE_X_SENDFILE type(bool) =  False
  VAR type(str) =  "/home/localhost/joelb/.lorax/0.94/var"
  VERSION type(str) =  "0.94.64"
======================= ===== ============================================================================

Manual Configuration
--------------------
Configuring lorax uses the ``lorax config`` command, run in the proper
environment using the lorax_env script.  The path to this script is
the value of LORAX_ROOT you chose at install time::

        /path/to/lorax_env -i


You will now get a ``lorax_env>`` prompt and you are ready to configure.
For example::

        lorax config host MY_IP_ADDRESS
        lorax config crashmail_email MY_EMAIL_ADDRESS
        lorax config secret_key # prints value of secret_key
        lorax create_instance

Now exit the lorax_env shell with control-D.  Run lorax and its associated
processes::

        /path/to/lorax_env -v start

The start command, when run with the ``-v`` switch should return a list of
started processes, all with status RUNNING.


Configuring by Script
---------------------
If you did a direct installation using the ``lorax_tool`` script, you should
review and edit the ``my_config.sh`` script to reflect the settings you wish
to use use for your installation.  Then run ``./lorax_tool configure_pkg`` to do the build.


INSTALL.configure+run.txt
We now configure lorax.  To begin, we need to enter the proper
environment, using the lorax_env script.  The path to this script is
the value of LORAX_ROOT you chose at install time:

	/path/to/lorax_env -i

You will now get a "lorax_env> " prompt and you are ready to configure.

configuration variables.  If you wish to change any of them, do it before
issuing the "configure_instance" command, which will write the startup
configurations which include such things as port numbers and which processes
to start.  Configuration may be done either via the "lorax config" command
or via environmental variables.  In particular the "LORAX_MODE" environmental
variable controls the setting of multiple variables in defining which processes
will get started.  Usually you will want to configure the following variables
in order to expose the lorax server to the internet (substituting the
desired values in the places beginning with MY_):

	lorax config host MY_IP_ADDRESS
	lorax config crashmail_email MY_EMAIL_ADDRESS
	lorax create_instance


Now it is time to run lorax and associate processes:

	supervisord; sleep 10
	supervisorctl status

The first command daemonizes lorax and associated processes.  The last
should return a list of commands, all with status RUNNING. You may then
enter "control-D" to exit the lorax_env environment.

Next, follow the instructions for testing in INSTALL.test.txt.


CHANGELOG.rst
Changelog
---------

.. snip

- Name of package abstracted away as much as possible to enable
  code reuse.

- Changed ``var/`` and ``etc/`` directory structures to play nicely
  with root at /.

- ``LORAX_ROOT`` is now always the location of the ``lorax_env``
  script and is defined at install time.

- ``run_in_lorax_env`` renamed to simply ``lorax_env``


0.94
~~~~

- ``run_in_lorax_env`` script created to define the two required
  environmental variables and activate the virtual environment
  from either Anaconda or normal versions.

- ``configure_instance`` command defined to configure the supervisord
  scripts.

- Supervisord is now used to start lorax and associated version.  
  supervisord>4.0 (currently only from git) is required due to python3
  requirement.

- Setuptools >30 is now required, due to moving many things into setup.cfg.

- FastTree source now included in distribution; FastTree binary will be built
  and put in virtual environment under name ``FastTree-lorax``.

- License is now 3-clause BSD (was 4-clause BSD).

-  ``config_value`` command is now simply ``config``.

0.93.2
~~~~~~

-  Created the ``create_anaconda_venv.sh`` script for downstream to use.

-  Changed the development platform to Anaconda python 3.6 in a virtualenv.

-  Changed this file to ReStructuredText.

0.93.1
~~~~~~

-   Code passes PEP8.

0.93.2
~~~~~~

-  Redirected stderr on HMM check call.

-  ``LORAX_INSTANCE_DIR`` environmental variable allows overriding the location of the instance
   configuration file.  Useful for installations where the default instance path is not
   writable by the user.

-  ``config_value`` with no arguments now indicates whether variables are set from instance
   config files or environmental variables.

-  ``config_value --verbose VAR`` gives verbose information about the type of the variable
   and where it was set.

- The default type for ``config_value`` is now, in priority order:
     * The type specified with the ``--vartype`` option, if it exists.
     * The type of the extant variable.
     * str if not previously defined and ``--vartype`` is not specified.

-  ``config_value --type`` renamed to ``--vartype``, and now accepts any valid python type
   for its argument.  Types other than str and bool are parsed via JSON and have JSON
   typing rules (including demanding use of double quotes in places).  Note that for
   complex types, you may have to enclose the argument in single quotes to keep the
   shell from mangling it.


-  `config_value VAR` now prints the value.

v0.92
~~~~~

-  `get_config` and `set_config` commands are now condensed to `config_value`
   which does set if ``VALUE`` is defined, set get if now.  If ``VAR`` is not
   defined, ``config_value`` lists all variables along with whether they were
   set from environmental variables or not.

-- Broke up ``PATH`` from a dictionary to be three simple key:value objects:
     * DATA_PATH
     * LOG_PATH
     * DIR_MODE

v0.91
~~~~~

-  Package data moved to proper level of hierarchy.

-  ``ALIGNMENT_QUEUE_TIMEOUT`` and ``TREE_QUEUE_TIMEOUT`` config parameters.  Was defaulting to
   180 seconds.

v0.9
~~~~

-  Logging format is now configurable.

-  Added commands for easier setup:
     * copy_test_files
     * get_config
     * set_config
     * test_logging

-  Added PhyD3 files to static and template folders.

-  Added ``GET`` method for ``/alignment`` target to return MSA's.

-  Created ``requirements.in`` and ``requirements.txt ``dependencies list via pip-sync.

-  Removed ete3 in favor of biopython due to ete3 dependency problems with BSD.
   Less control over Newick formatting, though...trees may not appear to be rooted
   in some downstream programs.


v0.8
~~~~

-  Changed the tree output to be rooted, entailing a change in the exact Newick format used.
   Root node names are family names.  Trees are now ladderized.  ete3 is now a dependency.

-  Added the ``/rq`` dashboard target. rq_dashboard is now a dependency.

-  Added commands.

-  Changed startup methods considerably.  See the startup scripts for details.

v0.7
~~~~

-  Names of executables are now configuration parameters.  This allows for overriding of executable
   names on platforms where the name may vary.

-  Added DELETE target for superfamiles.

-  Calculation methods now return JSON dictionaries of job listings that include the following:
      * ``id``: id string
      * ``description``: description string, e.g., "FastTree tree of superfamily aspartic_peptidases.myseqs"
      * ``status``: job.status, e.g., "deferred"
      * ``tasktype``: "align" or "tree"
      * ``taskname``: name of aligner or builder
      * ``family``: name of family
      * ``superfamily``: name of superfamily
      * ``is_queued``: True if queued.  This will not be true until after alignment in some cases.
      * ``is_started``: True if job is running.
      * ``is_finished``: True if job is finished (since queue was reset).
      * ``is_failed``: True if job failed due to an error.
      * ``created_at``: ISO format job creation time or "None".
      * ``enqueued_at``: ISO format time of queueing or "None".
      * ``ended_at``: ISO format ending time or "None".
      * ``started_at``: ISO format start time or "None".
      * ``estimated_job_time``: Estimated run time (wallclock) in seconds.
      * ``queue_name``: Name of queue (for future queries)
      * ``queue_position``: position in queue, if queued, or length of queue if not.
      * ``estimated_queue_time``: Estimated wallclock time of preceding jobs, in seconds.
   For now all estimated times are bogus and simply placeholders until timing models are established.

-  Config parameters are settable via environmental variables starting with ``"LORAX_"``.

-  Configuration parameters are now settable by instance as a pyfile.

-  Improved logfile content.

-  Removed ``/config`` target as insecure.


lorax/config.py
# -*- coding: utf-8 -*-
"""Define and handle configuration variables.
"""
#
# Library imports.
#
import datetime
#
# Name of this service.
#
SERVICE_NAME = 'lorax'
SERVICE_ORG = 'LegumeFederation'

def configure_app(app):
    """Flask app.config
    """
    #
    # Name of this service.
    #
    app.config['NAME'] = SERVICE_NAME
    #
    # Web site associated with this project.
    #
    app.config['PROJECT_HOME'] = 'https://github.com/' + SERVICE_ORG + '/' +SERVICE_NAME
    #
    # File path locations.  All of these are immutable except DATA.
    # Since different components run from different locations, these
    # must be absolute.  The immutable ones should be created before
    # runtime.
    #
    app.config['DATA'] = '/usr/local/var/data'
    #
    # Directory/file permissions.
    #
    app.config['DIR_MODE'] = '755'  # Note interaction with process umask
    #
    # RQ settings.  If "RQ_ASYNC" is False, then no queueing will be done.
    #
    app.config['TREE_QUEUE'] = 'treebuilding'
    app.config['ALIGNMENT_QUEUE'] = 'alignment'
    app.config['RQ_REDIS_URL'] = 'redis://redis:6379/0'
    app.config['RQ_SCHEDULER_QUEUE'] = app.config['ALIGNMENT_QUEUE']
    app.config['ALIGNMENT_QUEUE_TIMEOUT'] = 24 * 60 * 60  # 24 hours, in seconds
    app.config['TREE_QUEUE_TIMEOUT'] = 30 * 24 * 60 * 60  # 30 days, in seconds
    #
    # Definitions for alignment algorithms.
    #
    app.config['ALIGNERS'] = {
        'hmmalign': ["--trim", "--informat", "FASTA"]  # command-line arguments
    }
    app.config['HMMALIGN_EXE'] = 'hmmalign'
    #
    # Definitions for tree-building algorithms.
    #
    app.config['TREEBUILDERS'] = {
        "FastTree": {
            "peptide": ["-nopr", "-log", "peptide.log"],
            "DNA": ["-nt", "-gtr", "-log", "nucleotide.log", "-nopr"]
        },
        "RAxML": {
            "peptide": ["-b", "12345", "-p", "12345", "-N", "10", "-m",
                        "PROTGAMMABLOSUM62"],
            "DNA": ["-d"]
        }
    }
    #
    # Binaries.
    #
    app.config['FASTTREE_EXE'] = 'FastTree'
    app.config['RAXML_EXE'] = 'raxmlHPC'
    #
    # Current run.
    #
    app.config['DATETIME'] = datetime.datetime.now().strftime("%Y-%m-%d %H:%m:%S")
    #
    # Logging formatter.  Fields that are defined are:
    #    asctime: Time with too much precision
    #    levelname: Severity level.
    #    module:  module name.
    #    lineno: line number.
    #    pathname: File path.
    #    message: The message.
    #    url: Target address (if from a target).
    #    utcnow: Time in UTC.
    #    method: HTTP method.
    #    ip: Real IP address of the requester.
    #
    app.config['STDERR_LOG_FORMAT'] = '%(levelname)s: %(message)s'
